<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MAPITR Introduction -- Simulated Data • MAPITR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MAPITR Introduction -- Simulated Data">
<meta property="og:description" content="MAPITR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MAPITR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/MAPITR.Intro.SimulatedData.html">MAPITR Introduction -- Simulated Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mturchin20/MAPITR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>MAPITR Introduction – Simulated Data</h1>
                        <h4 class="author">Michael C Turchin</h4>
            
            <h4 class="date">2020-08-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mturchin20/MAPITR/blob/master/vignettes/MAPITR.Intro.SimulatedData.Rmd"><code>vignettes/MAPITR.Intro.SimulatedData.Rmd</code></a></small>
      <div class="hidden name"><code>MAPITR.Intro.SimulatedData.Rmd</code></div>

    </div>

    
    
<p>This is a simple introductory vignette to the <code>MAPITR</code> package. In this introductory example, we will use a simulated dataset to run <code>MAPITR</code> and explore some basic output. The dataset is provided with the package.</p>
<div id="running-mapitr" class="section level2">
<h2 class="hasAnchor">
<a href="#running-mapitr" class="anchor"></a>Running <code>MAPITR</code>
</h2>
<p>For this introductory vignette, we will be using <code>MAPITR_SimData_Phenotype</code>, <code>MAPITR_SimData_Genotypes</code>, and <code>MAPITR_SimData_Pathways</code>. These are simulated datasets created for the purposes of this vignette. To load <code>MAPITR</code> and the datasets, run the following:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"MAPITR"</span>)

<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">MAPITR_SimData_Phenotype</span>, <span class="no">MAPITR_SimData_Genotypes</span>, <span class="no">MAPITR_SimData_Pathways</span>, <span class="no">MAPITR_SimData_PCs</span>)</pre></body></html></div>
<p>First, we’ll take a look at the formats of our simulated input data:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">MAPITR_SimData_Genotypes</span>[<span class="fl">1</span>:<span class="fl">10</span>,<span class="fl">1</span>:<span class="fl">10</span>]</pre></body></html></div>
<pre><code>##    V1 V2 V3 V4 V5 V6 V7 V8 V9 V10
## 1   2  1  1  0  2  2  1  1  1   0
## 2   2  1  0  2  0  2  0  0  2   1
## 3   2  1  1  1  2  1  1  1  1   1
## 4   1  1  2  1  0  0  0  1  0   1
## 5   1  0  0  1  2  2  0  1  1   1
## 6   1  0  2  1  1  2  2  1  2   0
## 7   1  1  1  1  1  0  1  0  0   0
## 8   1  1  1  2  2  2  0  1  2   1
## 9   1  1  0  1  1  2  1  0  1   1
## 10  1  2  0  0  1  1  1  2  0   1</code></pre>
<p>Here you should see the first ten rows and columns of our genotype matrix. The genotype matrix is n x p, so the rows are our n individuals and the columns are our p SNPs. SNP genotypes, the entries of each column, are assumed to be 0, 1, or 2 – so note, diploid information is assumed to have been collapsed into a single value for each SNP. For example, the typical diploid setup of 0/0, 0/1, and 1/1, which represents the three possible combinations a given pair of reference (0) and alternative (1) alleles can have across both chromosomal copies in an individual, is now instead represented as 0 (0/0), 1 (0/1), or 2 (1/1). One way to derive a genotype matrix with this type of encoding is to use PLINK’s <code>--recodeAD</code> output option and remove all the <code>*_HET</code> columns afterwards.</p>
<p>Also note that the genotype matrix cannot have any missing genotypes (ie 0% rate of genotype missingness). This is necessary due to how the genetic relatedness matrices (GRMs), required components for the <code>MAPITR</code> variance component model, are constructed. For each entry in the genotype matrix, the only acceptable values are 0, 1, and 2.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">MAPITR_SimData_Phenotype</span>)</pre></body></html></div>
<pre><code>##           V1
## 1 -0.6796317
## 2 -1.4496919
## 3 -0.1318539
## 4 -1.4731807
## 5 -0.4276638
## 6  0.3917982</code></pre>
<p>Here you should see the first ten values of our input phenotype. <code>MAPITR</code> only runs on one phenotype at a time currently, so the main function only requires a single vector of phenotypic values as input.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">MAPITR_SimData_Pathways</span>)</pre></body></html></div>
<pre><code>##         V1
## 1 Pathway1
## 2 Pathway2
## 3 Pathway3
## 4 Pathway4
## 5 Pathway5
##                                                                                                                                                                                                                                                                                                      V2
## 1  293,648,355,675,17,158,754,502,49,212,62,187,529,594,258,497,964,305,132,774,117,12,269,460,521,276,360,317,834,806,669,655,74,894,311,104,244,580,470,216,309,624,739,24,963,449,283,661,202,252,384,268,433,761,286,699,620,762,832,428,68,603,330,213,125,938,315,179,591,266,226,663,804,387,396
## 2  668,90,201,739,866,642,667,916,357,743,781,348,204,21,481,29,444,658,676,532,548,509,822,500,949,24,331,253,218,197,967,746,757,656,206,352,782,768,939,522,982,452,696,981,475,161,926,765,602,940,664,615,109,57,241,764,551,99,785,216,897,513,243,261,812,327,622,740,778,210,50,867,726,889,562
## 3 114,254,521,18,960,983,929,809,849,405,504,697,673,766,298,804,617,530,824,17,732,933,551,398,292,308,108,213,838,490,435,315,752,922,861,920,583,524,150,79,668,261,822,443,196,395,709,434,578,374,584,601,761,864,754,295,989,758,496,935,317,777,604,477,485,774,540,939,159,54,375,460,365,2,846
## 4 632,279,429,915,96,863,448,18,826,510,957,963,859,11,930,821,686,890,950,380,864,698,599,312,169,756,575,876,482,921,355,413,426,704,115,962,208,250,394,502,692,571,418,42,527,901,657,782,992,592,452,330,85,41,639,428,282,289,319,816,469,347,944,674,918,382,348,558,573,290,328,862,416,650,860
## 5  250,507,491,451,312,720,666,138,92,714,843,702,427,136,755,973,917,309,76,65,116,368,816,586,379,860,812,388,743,543,795,911,369,296,584,939,758,897,741,174,645,330,269,21,276,510,82,132,478,292,206,552,969,417,655,12,351,142,172,797,573,506,339,210,53,319,943,457,674,777,425,739,729,840,535</code></pre>
<p>Here you should see a matrix with two columns. The first column should list the name of each pathway being analyzed. The second column should be a list of comma-separated column indices, representing the set of SNPs belonging to each associated pathway. These indices will be used by <code>MAPITR</code> to extract the proper SNPs per pathway from the input genotype matrix to conduct the marginal epistasis analysis. Note that the values for this second column are not the SNP IDs (eg rsID#s) or column headers (if your genotype matrix has them), they are the specific column locations in the input genotype matrix for each SNP.</p>
<p>Now to run <code>MAPITR</code>, we use the following code:</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">MAPITR_Output</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/MAPITR.html">MAPITR</a></span>(<span class="no">MAPITR_SimData_Genotypes</span>, <span class="no">MAPITR_SimData_Phenotype</span>, <span class="no">MAPITR_SimData_Pathways</span>)</pre></body></html></div>
<p>And to see the output <code><a href="../reference/MAPITR.html">MAPITR()</a></code> provides, we can run the following:</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">MAPITR_Output</span>$<span class="no">Results</span>)</pre></body></html></div>
<pre><code>##      Pathways      pValues       Est       PVE
## [1,]        1 0.0006793231 1.0684080 1.1281354
## [2,]        2 0.3746938563 0.2651890 0.2847803
## [3,]        3 0.2244317548 0.3696023 0.3918250
## [4,]        4 0.3775966094 0.2714509 0.2854367
## [5,]        5 0.4023264905 0.2529208 0.2742095</code></pre>
<p>The output shown from <code>MAPITR_Output$Results</code> should be a matrix with four columns. The first column should be the list of pathway names being analyzed (same order as input file), the second column should be the <code>MAPITR</code> p-values for each pathway, the third column should be the estimate of the epistatic GRM variance component (<code>G</code> in the <a href="http://biorxiv.org/">model</a>) for each pathway, and the fourth column should be the proportion of variance explained (PVE) for the phenotype by each pathway. Note that PVE estimates here are for pathways that generally contain overlapping variants – this impacts our interpretation of PVE. When single variant models are used in complex trait analysis, which is often the case, PVE can theoretically sum to 1 across all variants. A priori it is unclear what PVE should theoretically sum to across multiple pathways due to this common scenario where variants are shared across pathway definitions.</p>
<p>In this example, there should be 1 pathway that has a p-value below the Bonferonni-corrected threshold of .01 (.05 / 5, the number of pathways tested):</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">MAPITR_Output</span>$<span class="no">Results</span>[<span class="no">MAPITR_Output</span>$<span class="no">Results</span>[,<span class="fl">2</span>] <span class="kw">&lt;</span> <span class="fl">.01</span>,]</pre></body></html></div>
<pre><code>##     Pathways      pValues          Est          PVE 
## 1.0000000000 0.0006793231 1.0684079900 1.1281354388</code></pre>
<p>Also note this is a simulated dataset made to run quickly for this vignette – you will very likely need many more individuals and SNPs to identify an epistatic signal, and PVE should never be &gt; 1 in real data.</p>
<p>Lastly, we will go over how one can introduce a covariate to this test (on both sides of the model) if desired. However, we recommend that most covariates should be directly regressed out from the phenotype prior to <code>MAPITR</code> analysis. To include a covariate as part of <code>MAPITR</code> though, it is incorporated as part of the orthogonality projection matrix, H, used in equation 3 of the <a href="http://biorxiv.org/">MAPITR manuscript</a>. This projection matrix is applied to both sides of the model, meaning it is applied to both the phenotype <em>and</em> the genotypes, and rotates both spaces to be orthogonal to the components of H. By default a column of ones is included in this matrix to act as a y-intercept term. However, any additional covariates can be appended to this matrix. We currently choose to include top principal components as a part of this matrix.</p>
<p>To include the top 10 principal components of our simulated dataset into the <code>MAPITR</code> analysis, we can run the following:</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">MAPITR_Output2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/MAPITR.html">MAPITR</a></span>(<span class="no">MAPITR_SimData_Genotypes</span>, <span class="no">MAPITR_SimData_Phenotype</span>, <span class="no">MAPITR_SimData_Pathways</span>, <span class="kw">Covariates</span><span class="kw">=</span><span class="no">MAPITR_SimData_PCs</span>)
<span class="no">MAPITR_Output2</span>$<span class="no">Results</span>[<span class="no">MAPITR_Output2</span>$<span class="no">Results</span>[,<span class="fl">2</span>] <span class="kw">&lt;</span> <span class="fl">.01</span>,]</pre></body></html></div>
<pre><code>##     Pathways      pValues          Est          PVE 
## 1.0000000000 0.0001282748 1.1540812533 1.2908322801</code></pre>
<p>And after doing so, we find that including top principal components does further help refine our epistatic signal in this simulated dataset.</p>
<p>For further details on the <code>MAPITR</code> model, please see the associated <a href="http://biorxiv.org/">manuscript</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Michael Turchin, Lorin Crawford, Sohini Ramachandran.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
